src/docs/arch-btree.dox

```diff
@@ -54,34 +54,41 @@ the focus here will be on truncation for B-Tree data files (\c file: uri).
 
 @section btree_range_truncate_file Range Truncate On Files
 
+<!-- Caution: the long ref to the arch-fast-truncate page must stay on one line. -->
+
 To perform a truncate on a file, a start and stop cursor are positioned corresponding
 to the desired range provided by the user. The desired start and stop keys do not actually
 need to exist since the cursors are positioned by doing a search-near rather than a search.
 Once positioned, we do a page-by-page walk on the B-Tree, fast-truncating pages where
-possible. When a page is marked for fast-truncation, its \c WT_READ_TRUNCATE flag is set,
-and the code tries to delete the page without instantiating it into memory. If no errors
-are encountered, a page-deleted structure \c WT_PAGE_DELETED is allocated and initialized
-with the timestamp, transaction id and state information associated with the truncate.
-Finally, the page is published as deleted using the \c WT_REF_DELETED flag. Pages which are
-not eligible for fast truncation are pages where only part of it will be deleted (e.g. the
-first and last pages of the truncate range), pages with overflow items, or pages with
-prepared updates. These pages will have their records deleted individually.
+possible.
+See @ref ft_truncate for a detailed description, and see
+@ref arch-fast-truncate "the separate page on fast-truncate and deleted pages"
+for further description of other ways deleted pages appear and what happens to them after
+they are deleted.
 
 @section btree_truncate_interaction_with_other_operations Interaction With Other Operations
 
-Truncation differs from other operations in the sense that it does not have the same
-transactional semantics. If a range truncate is in progress and another transaction
-happens to insert a key into the same range, the behavior is not well-defined. A
-conflict may be detected, or both transaction may be permitted to commit. In the scenario
-that both commit, if the system crashes and recovery runs, the resulting state of the
-database may be different to what was in the cache had the crash not happened.
+Historically we have stated that truncation is not transactional: if a range truncate is in
+progress and another transaction happens to insert a key into the same range, the behavior is
+not well defined.
+This was a hedge for a known bug; that bug is now understood, and it is limited to cases in
+logged trees where the truncation passes \c NULL instead of a cursor for the start and/or end
+point and that point changes concurrently; it is possible in these circumstances for log
+replay to find and use a different position for the start or end of the tree than occurred in
+the original transaction.
```

Comment:
I agree that the problem applies to logged trees with recovery, but unless things have changed, I think it is broader than described above.

I think the problem is that you can truncate [A, C] in T1 and insert B in a concurrent T2. Depending on how T1 and T2 interleave, you could have the following outcomes:

1. T2 inserts B, T1 visits page containing B and gets a write conflict;
2. T1 goes past B, T2 inserts B, T1 commits, T2 commits — no conflict during execution, B successfully inserted
3. T1 goes past B, T2 inserts B, T2 commits, T1 commits — no conflict during execution, B successfully inserted, but it will be silently removed during recovery when T2 is applied after T1.

Since applications can’t readily distinguish between cases 2 and 3 (and the state of 3 after recovery is different to the original execution), the documentation tries to say “don’t do that”.

I’m not sure what’s special about the case where A=NULL (or presumably C=NULL?) — I always thought the issue was that recovery could produce a different result than was seen originally.

Reply:
If it's an insert and not an update, (1) can't happen (the new value isn't visible to the truncate if they're running concurrently) but the point about recovery order is a good one that I'd previously missed. That is: yes, because truncate only logs the range, write phantoms might disappear after recovery. I'll add that to the notes in WT-4158.

What's special about NULL is that we log NULL instead of resolving the endpoint and logging that; recovery then figures out the end of the table for itself and it's possible for it to get a different answer. I discovered this behavior when debugging WT-4158 a while back.

(And I'll update the text here)

Reply:
I have improved what's posted in WT-4158. The text here is fixed in [02cbfa9](https://github.com/wiredtiger/wiredtiger/pull/8237/commits/02cbfa981bf3b19022a5d8fcd84491bb833b3d2f).